src/corecel/data/LdgIterator.hh

```diff
@@ -0,0 +1,210 @@
+    CELER_CONSTEXPR_FUNCTION self& operator--() noexcept
+    {
+        --ptr_;
+        return *this;
+    }
+    CELER_CONSTEXPR_FUNCTION self operator--(int) noexcept
+    {
+        self tmp{ptr_};
+        --ptr_;
+        return tmp;
+    }
+
+    // RandomAccessIterator requirements
+    CELER_CONSTEXPR_FUNCTION self& operator+=(const difference_type n) noexcept
+    {
+        ptr_ += n;
+        return *this;
+    }
+    CELER_CONSTEXPR_FUNCTION self operator+(const difference_type n) const noexcept
+    {
+        self tmp{ptr_};
+        return tmp += n;
+    }
+    CELER_CONSTEXPR_FUNCTION self& operator-=(const difference_type n) noexcept
+    {
+        ptr_ -= n;
+        return *this;
+    }
+    CELER_CONSTEXPR_FUNCTION self operator-(const difference_type n) const noexcept
+    {
+        self tmp{ptr_};
+        return tmp -= n;
+    }
+    CELER_CONSTEXPR_FUNCTION difference_type
+    operator-(self const& it) const noexcept
+    {
+        return it.ptr_ - ptr_;
+    }
+    CELER_CONSTEXPR_FUNCTION reference
+    operator[](const difference_type n) const noexcept
+    {
+#if CELER_DEVICE_COMPILE
+        return __ldg(ptr_ + n);
+#else
+        return ptr_[n];
+#endif
+    }
+    CELER_CONSTEXPR_FUNCTION friend bool
+    operator<(self const& lhs, self const& rhs)
+    {
+        return lhs.ptr_ < rhs.ptr_;
+    }
+
+    // Conversion operators
+    CELER_CONSTEXPR_FUNCTION explicit operator pointer() const noexcept
+    {
+        return ptr_;
+    }
+    CELER_CONSTEXPR_FUNCTION explicit operator bool() const noexcept
+    {
+        return ptr_ != nullptr;
+    }
+
+  private:
+    pointer ptr_;
+};
+
+//---------------------------------------------------------------------------//
+// FREE FUNCTIONS
+//---------------------------------------------------------------------------//
+
+// RandomAccessIterator requirements
+template<class T>
+CELER_CONSTEXPR_FUNCTION bool
+operator>(LDGIterator<T> const& lhs, LDGIterator<T> const& rhs)
+{
+    return rhs < lhs;
+}
+template<class T>
+CELER_CONSTEXPR_FUNCTION bool
+operator<=(LDGIterator<T> const& lhs, LDGIterator<T> const& rhs)
+{
+    return !(lhs > rhs);
+}
+template<class T>
+CELER_CONSTEXPR_FUNCTION bool
+operator>=(LDGIterator<T> const& lhs, LDGIterator<T> const& rhs)
+{
+    return !(lhs < rhs);
+}
+template<class T>
+CELER_CONSTEXPR_FUNCTION LDGIterator<T>
+operator+(const typename LDGIterator<T>::difference_type n,
+          LDGIterator<T> const& it)
+{
+    return it + n;
+}
+
+// ForwardIterator requirements
+template<class T>
+CELER_CONSTEXPR_FUNCTION bool
+operator!=(LDGIterator<T> const& lhs, LDGIterator<T> const& rhs)
+{
+    return !(lhs == rhs);
+}
+
+// Iterator requirements
+template<class T>
+CELER_CONSTEXPR_FUNCTION void
+swap(LDGIterator<T>& lhs, LDGIterator<T>& rhs) noexcept
+{
+    return lhs.swap(rhs);
+}
+
+// Helper
+template<class T>
+inline LDGIterator<T> make_ldgiterator(T const* ptr) noexcept
+{
+    return LDGIterator<T>{ptr};
+}
```

Comment:
I don't think we need this with c++17; make a deduction guide?

Reply:
We probably don't even need the deduction guide since we already have a constructor taking the same arguments, this was mostly to be consistent with other classes such as `make_observer`or `make_span`.

Reply:
Oh yeah, I forget about the circumstances when CTAD works automatically... also because it's changed from C++17 to 20. I think one key one is that 17 can't deduce automatically for struct-like member initialization ...